
{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ORF: Open Reanding Frames\n",
    "\n",
    "### Problem  \n",
    "https://rosalind.info/problems/orf/\n",
    "\n",
    "Either strand of a DNA double helix can serve as the coding strand for RNA transcription. Hence, a given DNA string implies six total reading frames, or ways in which the same region of DNA can be translated into amino acids: three reading frames result from reading the string itself, whereas three more result from reading its reverse complement.\n",
    "\n",
    "An open reading frame (ORF) is one which starts from the start codon and ends by stop codon, without any other stop codons in between. Thus, a candidate protein string is derived by translating an open reading frame into amino acids until a stop codon is reached.\n",
    "\n",
    "**Given:** A DNA string s of length at most 1 kbp in FASTA format.\n",
    "\n",
    "**Return:** Every distinct candidate protein string that can be translated from ORFs of s. Strings can be returned in any order."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Level 1: Solve Rosalind problem ORF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Strategy:\n",
    "Not going for the most straight-forward path because I want to create useful intermediate steps and I want to retain frame information.\n",
    "1. Create dictionary with codons of all 6 possible frames (sense_1, sense_2, sense_3, asense_1, asense_2, asense_3).\n",
    "2. Find all start and stop codon positions in frames dictionary and store positions in new dictionary (still organized by frame).\n",
    "3. Select all the ORF's (i.e. codons from start to stop w/o intermediate stop codons), translate into peptide sequence and add peptide to new dictionary organized by frame.\n",
    "4. Print all the unique possible peptides.\n",
    "\n",
    "NOTE:\n",
    "Because I want to also practice bipython, I will create an alternative version using it whenever it is useful."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "codon_table = {'UUU':'F', 'UUC':'F', 'UUA':'L', 'UUG':'L', 'UCU':'S',\n",
    "               'UCC':'S', 'UCA':'S', 'UCG':'S', 'UAU':'Y', 'UAC':'Y',\n",
    "               'UAA':'stop', 'UAG':'stop', 'UGU':'C', 'UGC':'C', 'UGA':'stop',\n",
    "               'UGG':'W', 'CUU':'L', 'CUC':'L', 'CUA':'L', 'CUG':'L',\n",
    "               'CCU':'P', 'CCC':'P', 'CCA':'P', 'CCG':'P', 'CAU':'H',\n",
    "               'CAC':'H', 'CAA':'Q', 'CAG':'Q', 'CGU':'R', 'CGC':'R',\n",
    "               'CGA':'R', 'CGG':'R', 'AUU':'I', 'AUC':'I', 'AUA':'I',\n",
    "               'AUG':'M', 'ACU':'T', 'ACC':'T', 'ACA':'T', 'ACG':'T',\n",
    "               'AAU':'N', 'AAC':'N', 'AAA':'K', 'AAG':'K', 'AGU':'S',\n",
    "               'AGC':'S', 'AGA':'R', 'AGG':'R', 'GUU':'V', 'GUC':'V',\n",
    "               'GUA':'V', 'GUG':'V', 'GCU':'A', 'GCC':'A', 'GCA':'A',\n",
    "               'GCG':'A', 'GAU':'D', 'GAC':'D', 'GAA':'E', 'GAG':'E',\n",
    "               'GGU':'G', 'GGC':'G', 'GGA':'G', 'GGG':'G'}\n",
    "\n",
    "with open('rosalind_orf.txt','r') as file:\n",
    "    infile = file.read().split('\\n')\n",
    "    DNA = ''.join(infile[1:])\n",
    "    RNA_sense = DNA.replace('T','U')\n",
    "    RNA_asense = RNA_sense.replace('A','u').replace('U','a').replace('C','g').replace('G','c').upper()[::-1]\n",
    "    RNA= {'RNA_sense' : RNA_sense,\n",
    "          'RNA_asense' : RNA_asense}\n",
    "\n",
    "\n",
    "# 1. Create dictionary with codons of all 6 possible frames:\n",
    "\n",
    "frames = {}\n",
    "for strand, seq in RNA.items():\n",
    "    for frame_start in range(3):\n",
    "        key = f'{strand}_{frame_start+1}'\n",
    "        frames[key] = [seq[i:i+3] for i in range(frame_start,len(seq)-2, 3)]\n",
    "\n",
    "\n",
    "# 2. Find all start and stop codon positions in frames dictionary and store them in new dictionaries.\n",
    "# Note: could have also stored in list, but I wanted to retain the frame information for myself\n",
    "\n",
    "start_codon = 'AUG'\n",
    "stop_codon = {'UAA','UAG','UGA'}\n",
    "\n",
    "\n",
    "start_i = {}\n",
    "stop_i = {}\n",
    "\n",
    "for frame, codons in frames.items():\n",
    "    start_i[frame]= [i for i, codon in enumerate(codons) if codon == start_codon]\n",
    "    stop_i[frame]= [i for i, codon in enumerate(codons) if codon in stop_codon]\n",
    "\n",
    "\n",
    "# 3. Select all ORF's (start to stop codons w/o an intermediate stop codon), translate into peptide and organize by frame in new dictionary\n",
    "\n",
    "ORFs = {}\n",
    "for frame, codons in frames.items():\n",
    "    starts = start_i[frame]\n",
    "    stops = stop_i[frame]\n",
    "\n",
    "    for start in starts:\n",
    "        for stop in stops:\n",
    "            if stop > start:\n",
    "                mid_stops = [mid_codon for mid_codon in stops if start < mid_codon < stop] # Find intermediate stop codons\n",
    "                if not mid_stops: # Proceed if there aren't intermediate stop codons\n",
    "                    orf_codons = codons[start:stop]\n",
    "                    peptide = ''.join(codon_table[codon] for codon in orf_codons)\n",
    "                    if frame not in ORFs:\n",
    "                        ORFs[frame] = []\n",
    "                    ORFs[frame].append(peptide)\n",
    "                    break\n",
    "\n",
    "\n",
    "# 4. Print all the unique possible peptides\n",
    "\n",
    "unique_peptides = set() # Kind of like a list but for unique values and unordered\n",
    "for peptide_list in ORFs.values():\n",
    "    unique_peptides.update(peptide_list)\n",
    "for unique_peptide in unique_peptides:\n",
    "    print(unique_peptide)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Biopython variant"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Leveling up\n",
    "\n",
    "I want to expand on this Rosalind problem by tweaking the goals into incremental levels of complexity - each level introducing new skills to learn and build on.\n",
    "\n",
    "- Level 1: Solve the problem\n",
    "- Level 2: Analyse putative ORF data and generate fasta files\n",
    "- Level 3: Scale to analysis of multiple sequences and run through bash\n",
    "- Level 4: Interface\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Level 2: Analyse putative ORF data and generate fasta files\n",
    "\n",
    "**Create program to assemble and analyse data on the putative ORFs and peptides for a DNA sequence.**\n",
    "\n",
    "This means generating a data frame including:\n",
    "- Unique putative ORF and peptide ID; Frame and Strand; Start and stop positions; ORF length and peptide size; Overlap information\n",
    "And plotting some data:\n",
    "- Size x position plot; Overlaps\n",
    "\n",
    "Additionally, program should be interactive and ask whether you want to generate fasta files containing the putative ORFs (in DNA and/or RNA) and/or the respective peptides, along with what sequence identifying information you would like to include.\n",
    "\n",
    "Rather than just returning a list of all the putative peptides translated from a DNA string, I want to create a program to assemble a data frame storing more useful information about the putative ORF's of a DNA sequence. The resulting data frame would include:\n",
    "- ID (generate an ID for each putative ORF: dnaID_orf_1, dnaID_orf_2...)\n",
    "- Frame (Sense +1, Sense +2, Sense +3, Antisense +1, Antisense +2, Antisense +3)\n",
    "- Start (position of first nucleotide in ORF)\n",
    "- Stop (position of last nucleotide in ORF)\n",
    "- Stop codon sequence (which of the three stop codons?)\n",
    "- ORF length (nucleotides in putative ORF)\n",
    "- Peptide size (number of aminoacids in putative peptide)\n",
    "- Overlap (does ORF overlap with another ORF?)\n",
    "\n",
    "Learning goals:\n",
    "- Intro do Pandas data analysis\n",
    "    - Creating data-frames\n",
    "    - Plotting data\n",
    "- Writing fasta files\n",
    "- Creating interactive functions\n",
    "\n",
    "Steps:\n",
    "- [x] Use Pandas to create data-frame\n",
    "- [x] Plot data\n",
    "- [ ] Write fasta files (DNA, RNA and protein)\n",
    "- [ ] Integrate code into interactive function"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2.1 Use Pandas to create data-frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Step 1: Prepare sequence for analysis\n",
    "with open('rosalind_orf.txt', 'r') as file:\n",
    "    infile = file.read().splitlines()\n",
    "    print(infile)\n",
    "    dna_id = infile[0].replace('>','')\n",
    "    dna_seq = ''.join(infile[1:])\n",
    "    print(dna_seq)\n",
    "    rna_seq_sense = dna_seq.replace('T','U')\n",
    "    rna_seq_asense = rna_seq_sense.replace('A','u').replace('U','a').replace('C','g').replace('G','c').upper()[::-1]\n",
    "    rna_seq = {'Sense' : rna_seq_sense, 'Antisense' : rna_seq_asense}\n",
    "    dna_dict = {'ID' : [dna_id], 'DNA sequence' : [dna_seq] }\n",
    "    dna_seq_asense = dna_seq.replace('A','t').replace('T','a').replace('C','g').replace('G','c').upper()[::-1]\n",
    "\n",
    "# Step 2: Set start and stop codons\n",
    "start_codon = ['AUG']\n",
    "stop_codon = ['UAA','UAG','UGA']\n",
    "\n",
    "\n",
    "# Step 3: Create each data variable as a list\n",
    "index = []\n",
    "start_codon_i = []\n",
    "stop_codon_i = []\n",
    "strand = []\n",
    "frame = []\n",
    "length = []\n",
    "peptide_size = []\n",
    "fromhere_dna = []\n",
    "tohere_dna = []\n",
    "start_dna = []\n",
    "stop_dna = []\n",
    "overlap = []\n",
    "\n",
    "\n",
    "# Step 4: Fill-up the data lists\n",
    "# Gather all the data on position, direction and size of putative ORFs\n",
    "for dir, seq in rna_seq.items():\n",
    "    for i in range(len(seq)-5): # -5 because I only want to find start codons that have enough subsequent sequence for a stop codon\n",
    "        if seq[i:i+3] in start_codon: # if you find a start codon...\n",
    "            for j in range(i+3 ,len(seq)-2, 3):\n",
    "                if seq[j:j+3] in stop_codon: # ...then find a stop codon, fill-out the variables: \n",
    "                    if dir == 'Sense':\n",
    "                        start_codon_i.append(i)\n",
    "                        stop_codon_i.append(j)\n",
    "                        fromhere_dna.append(i+1) # +1 to change to a 1-base count\n",
    "                        tohere_dna.append(j+3) # +3 to include stop codon bp\n",
    "                        start_dna.append(i+1)\n",
    "                        stop_dna.append(j+3)\n",
    "                        strand.append(dir)\n",
    "                        if (i+3) % 3 == 0:\n",
    "                            frame.append(f'{dir} +1')\n",
    "                        elif (i+2) % 3 == 0:\n",
    "                            frame.append(f'{dir} +2')\n",
    "                        elif (i+1) % 3 == 0:\n",
    "                            frame.append(f'{dir} +3')\n",
    "                        length.append(j-i)\n",
    "                        peptide_size.append((j-i)//3) # // needed to get integer instead of float?\n",
    "                        break\n",
    "                    elif dir == 'Antisense':\n",
    "                        start_codon_i.append(i)\n",
    "                        stop_codon_i.append(j)\n",
    "                        fromhere_dna.append(len(dna_seq)-j-2) # -2 to include whole stop codon\n",
    "                        tohere_dna.append(len(dna_seq)-i)\n",
    "                        start_dna.append(len(dna_seq)-i)\n",
    "                        stop_dna.append(len(dna_seq)-j-2)\n",
    "                        strand.append(dir)\n",
    "                        if (i+3) % 3 == 0:\n",
    "                            frame.append(f'{dir} +1')\n",
    "                        elif (i+2) % 3 == 0:\n",
    "                            frame.append(f'{dir} +2')\n",
    "                        elif (i+1) % 3 == 0:\n",
    "                            frame.append(f'{dir} +3')\n",
    "                        length.append(j-i)\n",
    "                        peptide_size.append((j-i)//3)\n",
    "                        break\n",
    "\n",
    "# Assign an indivdual ID to the putative ORFs\n",
    "orf_id = []\n",
    "for i in range(len(fromhere_dna)):\n",
    "    orf_id.append(f'{id}_orf_{i+1}')\n",
    "\n",
    "# Look for overlap between putative ORFs in the DNA duplex\n",
    "for k in range(len(orf_id)):\n",
    "    overlaps = []\n",
    "    for l in range(len(fromhere_dna)):\n",
    "        if k != l:\n",
    "            if fromhere_dna[k] < tohere_dna[l] and tohere_dna[k] > fromhere_dna[l]:\n",
    "                overlaps.append(orf_id[l])\n",
    "    if overlaps:\n",
    "        overlap.append(', '.join(overlaps))\n",
    "    else:\n",
    "        overlap.append('None')\n",
    "\n",
    "# Step 5: Assemble data variable lists into a dictionary, then into a data-frame using pandas\n",
    "orf_data = {\n",
    "    'ID' : orf_id,\n",
    "    'First bp' : fromhere_dna, # First bp of putative ORF in DNA duplex \n",
    "    'Last bp' : tohere_dna, # Last bp of putative ORF in DNA duplex\n",
    "    'Frame' : frame, # Strand and coding frame of putative ORF\n",
    "    'Strand' : strand, # Strand where putative ORF is\n",
    "    'Length' : length, # Length of putative strand\n",
    "    'Peptide size (aa)' : peptide_size,\n",
    "    'Start codon index' : start_codon_i, # Index of start codon of putative ORF in respective strand. SEE NOTE BELLOW.\n",
    "    'Stop codon index' : stop_codon_i, # Index of stop codon of putative ORF in respective strand. SEE NOTE BELLOW.\n",
    "    'Start in DNA' : start_dna, # Position of start codon first nt in DNA duplex\n",
    "    'Stop in DNA' : stop_dna, # Position of stop codon last nt in DNA duplex\n",
    "    'Overlap in DNA' : overlap # Whether putative ORF overlaps with any others in the DNA duplex\n",
    "}\n",
    "\n",
    "# NOTE:\n",
    "# The reason why I gathered seemingly redundant positional data:\n",
    "# 'First bp' and 'Last bp' just indicate that an ORF goes from A to B in the DNA duplex, irrespective of the strand it is encoded in\n",
    "# 'Start in DNA' and 'Stop in DNA' is similar but it retains information of the ORFs orientation - i.e.: if the ORF is encoded in the antisense strand 'Stop in DNA' value is lower than the 'Start in DNA' value\n",
    "# The first set of variables is useful to simply know putative ORFs position in the DNA duplex\n",
    "# The second set of variables is useful for plotting putative ORFs position along the DNA in their correct orientation\n",
    "\n",
    "import pandas as pd\n",
    "orf_df = pd.DataFrame(orf_data).sort_values(by='First bp') # turn dictionary into data-frame\n",
    "orf_df.to_csv('orf_df.csv') # save in csv file\n",
    "\n",
    "\n",
    "dna_df = pd.DataFrame(dna_dict)\n",
    "dna_df.to_csv('inputdna_df.csv')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2.2 Plot data\n",
    "\n",
    "Maybe at some point I will try to do it in python, but for now I wanted to do it in R.\n",
    "It would be a good way to warm up my R skills and get some analysis done before I dive into plotting with python.\n",
    "In the next block, I will try to do that R plot. Remember to change to the R kernel for it to work!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2.3 Create FASTA file\n",
    "\n",
    "1. Use dataframe to generate putative ORF, mRNA and peptide sequences\n",
    "2. Write interactive function to generate fasta files with these sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 1: Set codon table\n",
    "rna_codon_table = {\n",
    "    'AUG': 'M',  # Start codon (Methionine)\n",
    "    'UUU': 'F', 'UUC': 'F',  # Phenylalanine\n",
    "    'UUA': 'L', 'UUG': 'L',  # Leucine\n",
    "    'CUU': 'L', 'CUC': 'L', 'CUA': 'L', 'CUG': 'L',  # Leucine\n",
    "    'AUU': 'I', 'AUC': 'I', 'AUA': 'I',  # Isoleucine\n",
    "    'GUU': 'V', 'GUC': 'V', 'GUA': 'V', 'GUG': 'V',  # Valine\n",
    "    'UCU': 'S', 'UCC': 'S', 'UCA': 'S', 'UCG': 'S',  # Serine\n",
    "    'CCU': 'P', 'CCC': 'P', 'CCA': 'P', 'CCG': 'P',  # Proline\n",
    "    'ACU': 'T', 'ACC': 'T', 'ACA': 'T', 'ACG': 'T',  # Threonine\n",
    "    'GCU': 'A', 'GCC': 'A', 'GCA': 'A', 'GCG': 'A',  # Alanine\n",
    "    'UAU': 'Y', 'UAC': 'Y',  # Tyrosine\n",
    "    'CAU': 'H', 'CAC': 'H',  # Histidine\n",
    "    'CAA': 'Q', 'CAG': 'Q',  # Glutamine\n",
    "    'AAU': 'N', 'AAC': 'N',  # Asparagine\n",
    "    'AAA': 'K', 'AAG': 'K',  # Lysine\n",
    "    'GAU': 'D', 'GAC': 'D',  # Aspartic acid\n",
    "    'GAA': 'E', 'GAG': 'E',  # Glutamic acid\n",
    "    'UGU': 'C', 'UGC': 'C',  # Cysteine\n",
    "    'UGG': 'W',  # Tryptophan\n",
    "    'CGU': 'R', 'CGC': 'R', 'CGA': 'R', 'CGG': 'R',  # Arginine\n",
    "    'AGU': 'S', 'AGC': 'S',  # Serine\n",
    "    'AGA': 'R', 'AGG': 'R',  # Arginine\n",
    "    'GGU': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G',  # Glycine\n",
    "    'UAA': 'Stop', 'UAG': 'Stop', 'UGA': 'Stop'  # Stop codons\n",
    "}\n",
    "\n",
    "# Step 2: Create lists of putative ORFs, mRNAs and peptides\n",
    "\n",
    "putative_orfs = []\n",
    "putative_mrnas = []\n",
    "putative_peptides = []\n",
    "\n",
    "for index, row in orf_df.iterrows():\n",
    "    if row['Strand'] == 'Sense':\n",
    "        putative_orfs.append(dna_seq[row['Start codon index']:(row['Stop codon index']+3)])\n",
    "    elif row['Strand'] == 'Antisense':\n",
    "        putative_orfs.append(dna_seq_asense[row['Start codon index']:(row['Stop codon index']+3)]) # To include the stop codon\n",
    "\n",
    "for index, row in orf_df.iterrows():\n",
    "    if row['Strand'] == 'Sense':\n",
    "        putative_mrnas.append(rna_seq_sense[row['Start codon index']:row['Stop codon index']])\n",
    "    elif row['Strand'] == 'Antisense':\n",
    "        putative_mrnas.append(rna_seq_asense[row['Start codon index']:row['Stop codon index']])\n",
    "\n",
    "for mrna in putative_mrnas:\n",
    "    peptide = ''\n",
    "    for codon_start in range(0, len(mrna), 3):\n",
    "        codon = mrna[codon_start:codon_start+3]\n",
    "        if codon in rna_codon_table:\n",
    "            aa = rna_codon_table.get(codon)\n",
    "            peptide += aa\n",
    "        else:\n",
    "            print('something is wrong')\n",
    "    putative_peptides.append(peptide)\n",
    "\n",
    "# Print the unique putative peptides (Rosalind wants it that way)\n",
    "# unique_putative_peptides = set(putative_peptides)\n",
    "#for peptide in unique_putative_peptides:\n",
    "    #print(peptide)\n",
    "\n",
    "# Step 3: Add sequence lists to the dataframe\n",
    "\n",
    "orf_df_seq = orf_df\n",
    "orf_df_seq['DNA'] = putative_orfs\n",
    "orf_df_seq['RNA'] = putative_mrnas\n",
    "orf_df_seq['Peptide'] = putative_peptides\n",
    "\n",
    "orf_df_seq.to_csv('orf_df_seq.csv') # save in csv file\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 4: Create interactive function to generate fasta files with putative ORF sequences (DNA, RNA or peptide)\n",
    "\n",
    "def fasta_generator(df):\n",
    "    valid_types = ['DNA', 'RNA', 'PEPTIDE']\n",
    "    sequence_type = input(f\"Which sequence type you would like for the fasta file?\\n{valid_types}\").strip().upper()\n",
    "    if sequence_type not in valid_types:\n",
    "        print(f\"Invalid choice. Please choose from {valid_types}.\")\n",
    "        return\n",
    "    df.columns = df.columns.str.upper()\n",
    "    fasta_file = f\"{dna_id}_putativeorfs_{sequence_type}.fasta\"\n",
    "    with open(fasta_file, 'w') as f:\n",
    "        for index, row in df.iterrows():\n",
    "            f.write(f\">{row['ID']}\\n\")\n",
    "            f.write(f\"{row[sequence_type]}\\n\")\n",
    "    print(f\"FASTA file '{fasta_file}' with putative {sequence_type} sequences has been created.\")\n",
    "\n",
    "fasta_generator(orf_df_seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Incorporating biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import packages\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "# PART 1: PREPARE DNA INPUT\n",
    "\n",
    "def fasta_dna_parser(fasta_file):\n",
    "    \"\"\"Parse FASTA file with DNA sequences into dataframe with DNA ID and sequence.\n",
    "    \n",
    "    Input: path to FASTA file\n",
    "    Output: pandas dataframe\"\"\"\n",
    "\n",
    "    # Create empty dictionary\n",
    "    dna_dict = {\n",
    "        'DNA ID': [],\n",
    "        'DNA Sequence': []\n",
    "    }\n",
    "\n",
    "    # Open and parse FASTA file, then store sequneces in dictionary\n",
    "    with open(fasta_file, 'r') as file:\n",
    "        for item in SeqIO.parse(file, 'fasta'):\n",
    "            dna_id = item.id\n",
    "            dna_seq = str(item.seq)\n",
    "            dna_dict['DNA ID'].append(dna_id)\n",
    "            dna_dict['DNA Sequence'].append(dna_seq)\n",
    "    \n",
    "    # Return dna sequences in dataframe         \n",
    "    return pd.DataFrame(dna_dict)\n",
    "\n",
    "\n",
    "# PART 2: Find putative ORFs on DNA sequences\n",
    "\n",
    "def orf_finder(dna_df):\n",
    "    \"\"\"Find putative ORFs in DNA sequences provided in dataframe,\n",
    "    and return dataframe with location data for putative ORFs and unique putative ORF ID.\n",
    "    Note that start and stop codon location is stored as 0-based python index of the first codon nt.\n",
    "    \n",
    "    Input: dataframe with DNA ID and sequences    \n",
    "    Output: dataframe with putative ORF location data\"\"\"\n",
    "\n",
    "    # Set start and stop codons\n",
    "    start_codon = ['ATG']\n",
    "    stop_codon = ['TAA', 'TAG', 'UGA']\n",
    "\n",
    "    # Create empty dictionary to store locations of putative ORFs\n",
    "    orf_dict = {\n",
    "        'Source DNA ID': [],\n",
    "        'Putative ORF ID': [],\n",
    "        'Strand': [],\n",
    "        'Start codon index': [],\n",
    "        'Stop codon index': []\n",
    "    }\n",
    "\n",
    "    # Prepare DNA from dataframe \n",
    "    for _, row in dna_df.iterrows():\n",
    "        dna_id = row['DNA ID']\n",
    "        dna_seq_sense = row['DNA Sequence']\n",
    "        dna_seq_asense = str(Seq(dna_seq_sense).reverse_complement())\n",
    "\n",
    "        # Search for putative ORFs and fill-out dictionary with data of found ORFs\n",
    "        for dir, seq in {'Sense': dna_seq_sense, 'Antisense': dna_seq_asense}.items():\n",
    "            for i in range(len(seq) - 5): \n",
    "                if seq[i:i+3] in start_codon:\n",
    "                    for j in range(i+3, len(seq)-2, 3):\n",
    "                        if seq[j:j+3] in stop_codon:\n",
    "                            orf_id = f'{dna_id.lower()}_orf_{len(orf_dict['Putative ORF ID'])+1}'\n",
    "                            orf_dict['Source DNA ID'].append(dna_id)\n",
    "                            orf_dict['Putative ORF ID'].append(orf_id)\n",
    "                            orf_dict['Strand'].append(dir)\n",
    "                            orf_dict['Start codon index'].append(i)\n",
    "                            orf_dict['Stop codon index'].append(j)\n",
    "                            break\n",
    "    \n",
    "    # Return dataframe with location data of putative ORFs\n",
    "    return pd.DataFrame(orf_dict)\n",
    "\n",
    "# PART 3: ASSEMBLE DATAFRAME WITH MORE PUTATIVE ORF DATA\n",
    "\n",
    "def orf_data_collector(orf_locations_df):\n",
    "    \"\"\"Gather more data regarding putative ORFs and store it in more comprehensive dataframe.\n",
    "    This data can be useful for subsequent analysis and plotting.\n",
    "    Examples of data to collect:\n",
    "    Location of first and last bp of putative ORF in the DNA duplex (First bp and Last bp),\n",
    "    an alternative location metric similar to the aforementioned but retaining orientation of the ORF (Start and Stop),\n",
    "    length of the mRNA (Length of mRNA), size of the peptide (Peptide size),\n",
    "    whether putative ORF overlaps with others in the DNA duplex (Overlap in DNA)...\n",
    "\n",
    "    Input: dataframe with locations (0-based index) of putative ORFs\n",
    "    Output: comprehensive dataframe with more data on putative ORFs\"\"\"\n",
    "    return\n",
    "    \n",
    "\n",
    "# PART 4: GENERATE FASTA FILES WITH PUTATIVE ORFS\n",
    "\n",
    "def orf_dna_fasta_generator(orf_df):\n",
    "    \"\"\"Write DNA FASTA files for putative ORFs found for each input DNA sequence. One file is created per source DNA.\n",
    "    Files contains the putative ORF DNA sequences, along with ORF ID, frame, length and overlap information.\"\"\"\n",
    "    return\n",
    "\n",
    "def orf_mrna_fasta_generator(orf_df):\n",
    "    \"\"\"Write RNA FASTA files for putative ORFs found for each input DNA sequence. One file is created per source DNA.\n",
    "    File contains the putative ORF mRNA sequences, along with ORF ID and length\"\"\"\n",
    "    return\n",
    "\n",
    "def orf_mrna_fasta_generator(orf_df):\n",
    "    \"\"\"Write peptide FASTA files for putative ORFs found for each input DNA sequence. One file is created per source DNA.\n",
    "    File contains the putative ORF peptide sequences, along with ORF ID and size\"\"\"\n",
    "    return\n",
    "\n",
    "\n",
    "\n",
    "#ros_dna_df = fasta_dna_parser('rosalind_orf.txt')\n",
    "#ros_orf_locations_df = orf_finder(ros_dna_df)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
