{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ros-orf level 1\n",
    "\n",
    "### Solve Rosalind problem\n",
    "\n",
    "https://rosalind.info/problems/orf/\n",
    "\n",
    "Either strand of a DNA double helix can serve as the coding strand for RNA transcription. Hence, a given DNA string implies six total reading frames, or ways in which the same region of DNA can be translated into amino acids: three reading frames result from reading the string itself, whereas three more result from reading its reverse complement.\n",
    "\n",
    "An open reading frame (ORF) is one which starts from the start codon and ends by stop codon, without any other stop codons in between. Thus, a candidate protein string is derived by translating an open reading frame into amino acids until a stop codon is reached.\n",
    "\n",
    "**Given:** A DNA string s of length at most 1 kbp in FASTA format.\n",
    "\n",
    "**Return:** Every distinct candidate protein string that can be translated from ORFs of s. Strings can be returned in any order."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Strategy:\n",
    "Not going for the most straight-forward path because I want to retain frame information.\n",
    "1. Create dictionary with codons of all 6 possible frames (sense_1, sense_2, sense_3, asense_1, asense_2, asense_3).\n",
    "2. Find all start and stop codon positions in frames dictionary and store positions in new dictionary (still organized by frame).\n",
    "3. Select all the ORF's (i.e. codons from start to stop w/o intermediate stop codons), translate into peptide sequence and add peptide to new dictionary organized by frame.\n",
    "4. Print all the unique possible peptides.\n",
    "\n",
    "NOTE:\n",
    "Because I want to also practice bipython, I will create an alternative version using it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGCCAUGUAGCUAACUCAGGUUACAUGGGGAUGACCCCGCGACUUGGAUUAGAGUCUCUUUUGGAAUAAGCCUGAAUGAUCCGAGUAGCAUCUCAG\n",
      "CUGAGAUGCUACUCGGAUCAUUCAGGCUUAUUCCAAAAGAGACUCUAAUCCAAGUCGCGGGGUCAUCCCCAUGUAACCUGAGUUAGCUACAUGGCU\n",
      "MLLGSFRLIPKETLIQVAGSSPCNLS\n",
      "MGMTPRLGLESLLE\n",
      "MTPRLGLESLLE\n",
      "M\n"
     ]
    }
   ],
   "source": [
    "codon_table = {\n",
    "    'AUG': 'M',  # Start codon (Methionine)\n",
    "    'UUU': 'F', 'UUC': 'F',  # Phenylalanine\n",
    "    'UUA': 'L', 'UUG': 'L',  # Leucine\n",
    "    'CUU': 'L', 'CUC': 'L', 'CUA': 'L', 'CUG': 'L',  # Leucine\n",
    "    'AUU': 'I', 'AUC': 'I', 'AUA': 'I',  # Isoleucine\n",
    "    'GUU': 'V', 'GUC': 'V', 'GUA': 'V', 'GUG': 'V',  # Valine\n",
    "    'UCU': 'S', 'UCC': 'S', 'UCA': 'S', 'UCG': 'S',  # Serine\n",
    "    'CCU': 'P', 'CCC': 'P', 'CCA': 'P', 'CCG': 'P',  # Proline\n",
    "    'ACU': 'T', 'ACC': 'T', 'ACA': 'T', 'ACG': 'T',  # Threonine\n",
    "    'GCU': 'A', 'GCC': 'A', 'GCA': 'A', 'GCG': 'A',  # Alanine\n",
    "    'UAU': 'Y', 'UAC': 'Y',  # Tyrosine\n",
    "    'CAU': 'H', 'CAC': 'H',  # Histidine\n",
    "    'CAA': 'Q', 'CAG': 'Q',  # Glutamine\n",
    "    'AAU': 'N', 'AAC': 'N',  # Asparagine\n",
    "    'AAA': 'K', 'AAG': 'K',  # Lysine\n",
    "    'GAU': 'D', 'GAC': 'D',  # Aspartic acid\n",
    "    'GAA': 'E', 'GAG': 'E',  # Glutamic acid\n",
    "    'UGU': 'C', 'UGC': 'C',  # Cysteine\n",
    "    'UGG': 'W',  # Tryptophan\n",
    "    'CGU': 'R', 'CGC': 'R', 'CGA': 'R', 'CGG': 'R',  # Arginine\n",
    "    'AGU': 'S', 'AGC': 'S',  # Serine\n",
    "    'AGA': 'R', 'AGG': 'R',  # Arginine\n",
    "    'GGU': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G',  # Glycine\n",
    "    'UAA': 'Stop', 'UAG': 'Stop', 'UGA': 'Stop'  # Stop codons\n",
    "}\n",
    "\n",
    "\n",
    "with open(\"rosalind_orf_sample.txt\", \"r\") as file:\n",
    "    infile = file.read().split(\"\\n\")\n",
    "    DNA = \"\".join(infile[1:])\n",
    "    RNA_sense = DNA.replace(\"T\", \"U\")\n",
    "    RNA_asense = (\n",
    "        RNA_sense\n",
    "        .replace(\"A\", \"u\")\n",
    "        .replace(\"U\", \"a\")\n",
    "        .replace(\"C\", \"g\")\n",
    "        .replace(\"G\", \"c\")\n",
    "        .upper()[::-1]\n",
    "    )\n",
    "    RNA = {\"RNA_sense\": RNA_sense, \"RNA_asense\": RNA_asense}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# 1. Create dictionary with codons of all 6 possible frames:\n",
    "\n",
    "frames = {}\n",
    "\n",
    "for strand, seq in RNA.items():\n",
    "    for frame_start in range(3):\n",
    "        key = f\"{strand}_{frame_start+1}\"\n",
    "        frames[key] = [seq[i : i + 3] for i in range(frame_start, len(seq) - 2, 3)]\n",
    "\n",
    "\n",
    "\n",
    "# 2. Find all start and stop codon positions in frames dictionary and store them in new dictionaries.\n",
    "\n",
    "start_codon = \"AUG\"\n",
    "stop_codon = {\"UAA\", \"UAG\", \"UGA\"}\n",
    "\n",
    "start_i = {}\n",
    "stop_i = {}\n",
    "\n",
    "\n",
    "for frame, codons in frames.items():\n",
    "    start_i[frame] = [i for i, codon in enumerate(codons) if codon == start_codon]\n",
    "    stop_i[frame] = [i for i, codon in enumerate(codons) if codon in stop_codon]\n",
    "\n",
    "\n",
    "# 3. Select all ORF's (start to stop codons w/o an intermediate stop codon), translate into peptide and organize by frame in new dictionary\n",
    "\n",
    "\n",
    "ORFs = {}\n",
    "\n",
    "for frame, codons in frames.items():\n",
    "    starts = start_i[frame]\n",
    "    stops = stop_i[frame]\n",
    "\n",
    "    for start in starts:\n",
    "        for stop in stops:\n",
    "            if stop > start:\n",
    "                mid_stops = [mid_codon for mid_codon in stops if start < mid_codon < stop]  # Find intermediate stop codons\n",
    "                if not mid_stops:  # Proceed if there aren't intermediate stop codons\n",
    "                    orf_codons = codons[start:stop]\n",
    "                    peptide = \"\".join(codon_table[codon] for codon in orf_codons)\n",
    "                    if frame not in ORFs:\n",
    "                        ORFs[frame] = []\n",
    "                    ORFs[frame].append(peptide)\n",
    "                    break\n",
    "\n",
    "\n",
    "\n",
    "# 4. Print all the distinct possible peptides\n",
    "\n",
    "unique_peptides = set()  # Kind of like a list but for unique values and unordered\n",
    "\n",
    "for peptide_list in ORFs.values():\n",
    "    unique_peptides.update(peptide_list)\n",
    "\n",
    "for unique_peptide in unique_peptides:\n",
    "    print(unique_peptide)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Biopython version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MLLGSFRLIPKETLIQVAGSSPCNLS\n",
      "MGMTPRLGLESLLE\n",
      "MTPRLGLESLLE\n",
      "M\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "\n",
    "with open(\"rosalind_orf_sample.txt\", \"r\") as file:\n",
    "    items = SeqIO.read(file, \"fasta\")\n",
    "    DNA = items.seq\n",
    "    RNA_sense = DNA.transcribe()\n",
    "    RNA_asense = RNA_sense.reverse_complement().transcribe()\n",
    "    RNA = {\"RNA_sense\": RNA_sense, \"RNA_asense\": RNA_asense}\n",
    "\n",
    "\n",
    "# 1. Create dictionary with codons of all 6 possible frames:\n",
    "\n",
    "frames = {}\n",
    "for strand, seq in RNA.items():\n",
    "    for frame_start in range(3):\n",
    "        key = f\"{strand}_{frame_start+1}\"\n",
    "        frames[key] = [str(seq[i : i + 3]) for i in range(frame_start, len(seq) - 2, 3)]\n",
    "\n",
    "\n",
    "# 2. Find all start and stop codons in frames dictionary, extract putative ORFs and store in new dictionary\n",
    "\n",
    "start_codon = \"AUG\"\n",
    "stop_codon = {\"UAA\", \"UAG\", \"UGA\"}\n",
    "\n",
    "ORFs = {}\n",
    "\n",
    "# Iterate over codons for each frame\n",
    "for frame, codons in frames.items():\n",
    "\n",
    "    # Add frame as key to ORFs dictionary\n",
    "    ORFs[frame] = []\n",
    "\n",
    "    # Store position of start and stop codons in two lists\n",
    "    # enumerate(codons) generates tuple (index, codon)\n",
    "    starts = [i for i, codon in enumerate(codons) if codon == start_codon]\n",
    "    stops = [i for i, codon in enumerate(codons) if codon in stop_codon]\n",
    "\n",
    "    # INTERMEDIATE REPORT:\n",
    "    # print(f\"\\nFrame: {frame}\")\n",
    "    # print(f\"Start codon positions: {starts}\")\n",
    "    # print(f\"Stop codon positions: {stops}\")\n",
    "\n",
    "    # Iterate over start and stop codons positions\n",
    "    for start in starts:\n",
    "        for stop in stops:\n",
    "            # If there is a start codon followed by stop codon...\n",
    "            if stop > start:\n",
    "                # with no stop codons in between,\n",
    "                mid_stops = [mid_stop for mid_stop in stops if start < mid_stop < stop]\n",
    "                if not mid_stops:  # you found a putative ORF!\n",
    "                    # Assemble the mRNA sequence, translate into peptide and store it in ORFs dictionary\n",
    "                    mrna = \"\".join(codons[start:stop])\n",
    "                    peptide = Seq(mrna).translate()\n",
    "                    ORFs[frame].append(str(peptide))\n",
    "                    break\n",
    "\n",
    "\n",
    "# 3. Print all the distinct peptides potentially encoded in the DNA sequence\n",
    "\n",
    "unique_peptides = set()  # Kind of like a list but for unique values and unordered\n",
    "for peptide_list in ORFs.values():\n",
    "    unique_peptides.update(peptide_list)\n",
    "for unique_peptide in unique_peptides:\n",
    "    print(unique_peptide)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Final thoughts\n",
    "\n",
    "Not a very efficient way to go about it. It works fine with small DNA sequences, but I imagine it would not deal well with large sequences. In the next step I will refrain from creating the frames dictionary.\n",
    "\n",
    "In the future I can also avoid doing the ORF search on RNA, could just do it in the DNA sequence.\n",
    "\n",
    "Does biopython make things easier?\n",
    "Maybe, once I know its quirks. It simplifies parsing the FASTA file, transcribing and translating.\n",
    "But I found it very annoying that `.reversecomplement()` turns the sequence back to DNA. It was a headache to realize that that was why the code was not working."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
