{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ORF: Open Reanding Frames\n",
    "\n",
    "### Problem  \n",
    "https://rosalind.info/problems/orf/\n",
    "\n",
    "Either strand of a DNA double helix can serve as the coding strand for RNA transcription. Hence, a given DNA string implies six total reading frames, or ways in which the same region of DNA can be translated into amino acids: three reading frames result from reading the string itself, whereas three more result from reading its reverse complement.\n",
    "\n",
    "An open reading frame (ORF) is one which starts from the start codon and ends by stop codon, without any other stop codons in between. Thus, a candidate protein string is derived by translating an open reading frame into amino acids until a stop codon is reached.\n",
    "\n",
    "**Given:** A DNA string s of length at most 1 kbp in FASTA format.\n",
    "\n",
    "**Return:** Every distinct candidate protein string that can be translated from ORFs of s. Strings can be returned in any order."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Level 1: Solve Rosalind problem ORF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Strategy:\n",
    "Not going for the most straight-forward path because I want to create useful intermediate steps and I want to retain frame information.\n",
    "1. Create dictionary with codons of all 6 possible frames (sense_1, sense_2, sense_3, asense_1, asense_2, asense_3).\n",
    "2. Find all start and stop codon positions in frames dictionary and store positions in new dictionary (still organized by frame).\n",
    "3. Select all the ORF's (i.e. codons from start to stop w/o intermediate stop codons), translate into peptide sequence and add peptide to new dictionary organized by frame.\n",
    "4. Print all the unique possible peptides.\n",
    "\n",
    "NOTE:\n",
    "Because I want to also practice bipython, I will create an alternative version using it whenever it is useful."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MVNRILDRLY\n",
      "MDCQDYLRLWSRLATGADFSTWNSLLSAAVVILQTSSIDKLPIYFYPRFL\n",
      "MISSIAKSSGFAIVRNRSCYSSKLSSSHISLG\n",
      "M\n",
      "MSPVL\n",
      "MGKQASFEGAIQRGWPHDTGS\n",
      "MRQTNLT\n",
      "MGKQASSFDSKLLDCLLDESGPIVGFTGTLDRNRLAAYLSN\n",
      "MSIQRPVS\n",
      "MTRGHETGR\n",
      "MLMEVTLGAALPTSVGSRIRTRLVISYSVGSLPSSCT\n",
      "MKRGAECSWKSL\n",
      "MTSRVLIRDPTEVGSAAPRVTSMSIQRPVS\n",
      "MRPTSLNRALE\n",
      "MDDNSRCN\n",
      "MTPVFP\n",
      "MPLLRHT\n",
      "MFFHRGML\n",
      "MDT\n",
      "MSHTVIRGFGHSDSSRPQGALGAGGLSSVLAPACASMPLLRHT\n",
      "ML\n",
      "MTKTSYYSVRHHNIPRWKNMDDNSRCN\n",
      "MPYGLPRLFTIVVPACYGGGLLHVEQSPLCCCCNLAN\n",
      "MTLSIAGQGFG\n",
      "MTPCHAANLAESRPRMTPVFP\n",
      "MMSHTVIRGFGHSDSSRPQGALGAGGLSSVLAPACASMPLLRHT\n",
      "MEVTLGAALPTSVGSRIRTRLVISYSVGSLPSSCT\n"
     ]
    }
   ],
   "source": [
    "codon_table = {'UUU':'F', 'UUC':'F', 'UUA':'L', 'UUG':'L', 'UCU':'S',\n",
    "               'UCC':'S', 'UCA':'S', 'UCG':'S', 'UAU':'Y', 'UAC':'Y',\n",
    "               'UAA':'stop', 'UAG':'stop', 'UGU':'C', 'UGC':'C', 'UGA':'stop',\n",
    "               'UGG':'W', 'CUU':'L', 'CUC':'L', 'CUA':'L', 'CUG':'L',\n",
    "               'CCU':'P', 'CCC':'P', 'CCA':'P', 'CCG':'P', 'CAU':'H',\n",
    "               'CAC':'H', 'CAA':'Q', 'CAG':'Q', 'CGU':'R', 'CGC':'R',\n",
    "               'CGA':'R', 'CGG':'R', 'AUU':'I', 'AUC':'I', 'AUA':'I',\n",
    "               'AUG':'M', 'ACU':'T', 'ACC':'T', 'ACA':'T', 'ACG':'T',\n",
    "               'AAU':'N', 'AAC':'N', 'AAA':'K', 'AAG':'K', 'AGU':'S',\n",
    "               'AGC':'S', 'AGA':'R', 'AGG':'R', 'GUU':'V', 'GUC':'V',\n",
    "               'GUA':'V', 'GUG':'V', 'GCU':'A', 'GCC':'A', 'GCA':'A',\n",
    "               'GCG':'A', 'GAU':'D', 'GAC':'D', 'GAA':'E', 'GAG':'E',\n",
    "               'GGU':'G', 'GGC':'G', 'GGA':'G', 'GGG':'G'}\n",
    "\n",
    "with open('rosalind_orf.txt','r') as file:\n",
    "    infile = file.read().split('\\n')\n",
    "    DNA = ''.join(infile[1:])\n",
    "    RNA_sense = DNA.replace('T','U')\n",
    "    RNA_asense = RNA_sense.replace('A','u').replace('U','a').replace('C','g').replace('G','c').upper()[::-1]\n",
    "    RNA= {'RNA_sense' : RNA_sense,\n",
    "          'RNA_asense' : RNA_asense}\n",
    "\n",
    "\n",
    "# 1. Create dictionary with codons of all 6 possible frames:\n",
    "\n",
    "frames = {}\n",
    "for strand, seq in RNA.items():\n",
    "    for frame_start in range(3):\n",
    "        key = f'{strand}_{frame_start+1}'\n",
    "        frames[key] = [seq[i:i+3] for i in range(frame_start,len(seq)-2, 3)]\n",
    "\n",
    "\n",
    "# 2. Find all start and stop codon positions in frames dictionary and store them in new dictionaries.\n",
    "# Note: could have also stored in list, but I wanted to retain the frame information for myself\n",
    "\n",
    "start_codon = 'AUG'\n",
    "stop_codon = {'UAA','UAG','UGA'}\n",
    "\n",
    "\n",
    "start_i = {}\n",
    "stop_i = {}\n",
    "\n",
    "for frame, codons in frames.items():\n",
    "    start_i[frame]= [i for i, codon in enumerate(codons) if codon == start_codon]\n",
    "    stop_i[frame]= [i for i, codon in enumerate(codons) if codon in stop_codon]\n",
    "\n",
    "\n",
    "# 3. Select all ORF's (start to stop codons w/o an intermediate stop codon), translate into peptide and organize by frame in new dictionary\n",
    "\n",
    "ORFs = {}\n",
    "for frame, codons in frames.items():\n",
    "    starts = start_i[frame]\n",
    "    stops = stop_i[frame]\n",
    "\n",
    "    for start in starts:\n",
    "        for stop in stops:\n",
    "            if stop > start:\n",
    "                mid_stops = [mid_codon for mid_codon in stops if start < mid_codon < stop] # Find intermediate stop codons\n",
    "                if not mid_stops: # Proceed if there aren't intermediate stop codons\n",
    "                    orf_codons = codons[start:stop]\n",
    "                    peptide = ''.join(codon_table[codon] for codon in orf_codons)\n",
    "                    if frame not in ORFs:\n",
    "                        ORFs[frame] = []\n",
    "                    ORFs[frame].append(peptide)\n",
    "                    break\n",
    "\n",
    "\n",
    "# 4. Print all the unique possible peptides\n",
    "\n",
    "unique_peptides = set() # Kind of like a list but for unique values and unordered\n",
    "for peptide_list in ORFs.values():\n",
    "    unique_peptides.update(peptide_list)\n",
    "for unique_peptide in unique_peptides:\n",
    "    print(unique_peptide)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Biopython variant"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Leveling up\n",
    "\n",
    "I want to expand on this Rosalind problem by tweaking the goals into incremental levels of complexity - each level introducing new skills to learn and build on.\n",
    "\n",
    "- Level 1: Solve the problem\n",
    "- Level 2: Analyse putative ORF data and generate fasta files\n",
    "- Level 3: Scale to analysis of multiple sequences and run through bash\n",
    "- Level 4: Interface\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Level 2: Analyse putative ORF data and generate fasta files\n",
    "\n",
    "**Create program to assemble and analyse data on the putative ORFs and peptides for a DNA sequence.**\n",
    "\n",
    "This means generating a data frame including:\n",
    "- Unique putative ORF and peptide ID; Frame and Strand; Start and stop positions; ORF length and peptide size; Overlap information\n",
    "And plotting some data:\n",
    "- Size x position plot; Overlaps\n",
    "\n",
    "Additionally, program should be interactive and ask whether you want to generate fasta files containing the putative ORFs (in DNA and/or RNA) and/or the respective peptides, along with what sequence identifying information you would like to include.\n",
    "\n",
    "Rather than just returning a list of all the putative peptides translated from a DNA string, I want to create a program to assemble a data frame storing more useful information about the putative ORF's of a DNA sequence. The resulting data frame would include:\n",
    "- ID (generate an ID for each putative ORF: dnaID_orf_1, dnaID_orf_2...)\n",
    "- Frame (Sense +1, Sense +2, Sense +3, Antisense +1, Antisense +2, Antisense +3)\n",
    "- Start (position of first nucleotide in ORF)\n",
    "- Stop (position of last nucleotide in ORF)\n",
    "- Stop codon sequence (which of the three stop codons?)\n",
    "- ORF length (nucleotides in putative ORF)\n",
    "- Peptide size (number of aminoacids in putative peptide)\n",
    "- Overlap (does ORF overlap with another ORF?)\n",
    "\n",
    "Learning goals:\n",
    "- Intro do Pandas data analysis\n",
    "    - Creating data-frames\n",
    "    - Plotting data\n",
    "- Writing fasta files\n",
    "- Creating interactive functions\n",
    "\n",
    "Steps:\n",
    "- [x] Use Pandas to create data-frame\n",
    "- [ ] Plot data\n",
    "- [ ] Write fasta files (DNA, RNA and protein)\n",
    "- [ ] Integrate code into interactive function"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2.1 Use Pandas to create data-frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "926\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Frame</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Length</th>\n",
       "      <th>Peptide size (aa)</th>\n",
       "      <th>Overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Rosalind_3903_orf_1</td>\n",
       "      <td>35</td>\n",
       "      <td>65</td>\n",
       "      <td>Sense +3</td>\n",
       "      <td>Sense</td>\n",
       "      <td>30</td>\n",
       "      <td>10</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Rosalind_3903_orf_2</td>\n",
       "      <td>198</td>\n",
       "      <td>279</td>\n",
       "      <td>Sense +1</td>\n",
       "      <td>Sense</td>\n",
       "      <td>81</td>\n",
       "      <td>27</td>\n",
       "      <td>Rosalind_3903_orf_2, Rosalind_3903_orf_2, Rosa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Rosalind_3903_orf_3</td>\n",
       "      <td>255</td>\n",
       "      <td>279</td>\n",
       "      <td>Sense +1</td>\n",
       "      <td>Sense</td>\n",
       "      <td>24</td>\n",
       "      <td>8</td>\n",
       "      <td>Rosalind_3903_orf_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Rosalind_3903_orf_4</td>\n",
       "      <td>308</td>\n",
       "      <td>398</td>\n",
       "      <td>Sense +3</td>\n",
       "      <td>Sense</td>\n",
       "      <td>90</td>\n",
       "      <td>30</td>\n",
       "      <td>Rosalind_3903_orf_4, Rosalind_3903_orf_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Rosalind_3903_orf_5</td>\n",
       "      <td>374</td>\n",
       "      <td>398</td>\n",
       "      <td>Sense +3</td>\n",
       "      <td>Sense</td>\n",
       "      <td>24</td>\n",
       "      <td>8</td>\n",
       "      <td>Rosalind_3903_orf_5, Rosalind_3903_orf_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Rosalind_3903_orf_6</td>\n",
       "      <td>397</td>\n",
       "      <td>460</td>\n",
       "      <td>Sense +2</td>\n",
       "      <td>Sense</td>\n",
       "      <td>63</td>\n",
       "      <td>21</td>\n",
       "      <td>Rosalind_3903_orf_6, Rosalind_3903_orf_6, Rosa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Rosalind_3903_orf_7</td>\n",
       "      <td>410</td>\n",
       "      <td>443</td>\n",
       "      <td>Sense +3</td>\n",
       "      <td>Sense</td>\n",
       "      <td>33</td>\n",
       "      <td>11</td>\n",
       "      <td>Rosalind_3903_orf_7, Rosalind_3903_orf_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Rosalind_3903_orf_8</td>\n",
       "      <td>442</td>\n",
       "      <td>460</td>\n",
       "      <td>Sense +2</td>\n",
       "      <td>Sense</td>\n",
       "      <td>18</td>\n",
       "      <td>6</td>\n",
       "      <td>Rosalind_3903_orf_8, Rosalind_3903_orf_8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Rosalind_3903_orf_9</td>\n",
       "      <td>465</td>\n",
       "      <td>588</td>\n",
       "      <td>Sense +1</td>\n",
       "      <td>Sense</td>\n",
       "      <td>123</td>\n",
       "      <td>41</td>\n",
       "      <td>Rosalind_3903_orf_9, Rosalind_3903_orf_9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Rosalind_3903_orf_10</td>\n",
       "      <td>517</td>\n",
       "      <td>532</td>\n",
       "      <td>Sense +2</td>\n",
       "      <td>Sense</td>\n",
       "      <td>15</td>\n",
       "      <td>5</td>\n",
       "      <td>Rosalind_3903_orf_10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Rosalind_3903_orf_11</td>\n",
       "      <td>774</td>\n",
       "      <td>870</td>\n",
       "      <td>Sense +1</td>\n",
       "      <td>Sense</td>\n",
       "      <td>96</td>\n",
       "      <td>32</td>\n",
       "      <td>Rosalind_3903_orf_11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Rosalind_3903_orf_12</td>\n",
       "      <td>860</td>\n",
       "      <td>857</td>\n",
       "      <td>Antisense +1</td>\n",
       "      <td>Antisense</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>Rosalind_3903_orf_12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Rosalind_3903_orf_13</td>\n",
       "      <td>776</td>\n",
       "      <td>767</td>\n",
       "      <td>Antisense +1</td>\n",
       "      <td>Antisense</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Rosalind_3903_orf_14</td>\n",
       "      <td>701</td>\n",
       "      <td>590</td>\n",
       "      <td>Antisense +1</td>\n",
       "      <td>Antisense</td>\n",
       "      <td>111</td>\n",
       "      <td>37</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Rosalind_3903_orf_15</td>\n",
       "      <td>694</td>\n",
       "      <td>544</td>\n",
       "      <td>Antisense +2</td>\n",
       "      <td>Antisense</td>\n",
       "      <td>150</td>\n",
       "      <td>50</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Rosalind_3903_orf_16</td>\n",
       "      <td>484</td>\n",
       "      <td>466</td>\n",
       "      <td>Antisense +2</td>\n",
       "      <td>Antisense</td>\n",
       "      <td>18</td>\n",
       "      <td>6</td>\n",
       "      <td>Rosalind_3903_orf_16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Rosalind_3903_orf_17</td>\n",
       "      <td>461</td>\n",
       "      <td>398</td>\n",
       "      <td>Antisense +1</td>\n",
       "      <td>Antisense</td>\n",
       "      <td>63</td>\n",
       "      <td>21</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Rosalind_3903_orf_18</td>\n",
       "      <td>412</td>\n",
       "      <td>385</td>\n",
       "      <td>Antisense +2</td>\n",
       "      <td>Antisense</td>\n",
       "      <td>27</td>\n",
       "      <td>9</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Rosalind_3903_orf_19</td>\n",
       "      <td>399</td>\n",
       "      <td>363</td>\n",
       "      <td>Antisense +3</td>\n",
       "      <td>Antisense</td>\n",
       "      <td>36</td>\n",
       "      <td>12</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Rosalind_3903_orf_20</td>\n",
       "      <td>382</td>\n",
       "      <td>271</td>\n",
       "      <td>Antisense +2</td>\n",
       "      <td>Antisense</td>\n",
       "      <td>111</td>\n",
       "      <td>37</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Rosalind_3903_orf_21</td>\n",
       "      <td>376</td>\n",
       "      <td>271</td>\n",
       "      <td>Antisense +2</td>\n",
       "      <td>Antisense</td>\n",
       "      <td>105</td>\n",
       "      <td>35</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Rosalind_3903_orf_22</td>\n",
       "      <td>257</td>\n",
       "      <td>233</td>\n",
       "      <td>Antisense +1</td>\n",
       "      <td>Antisense</td>\n",
       "      <td>24</td>\n",
       "      <td>8</td>\n",
       "      <td>Rosalind_3903_orf_22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Rosalind_3903_orf_23</td>\n",
       "      <td>239</td>\n",
       "      <td>233</td>\n",
       "      <td>Antisense +1</td>\n",
       "      <td>Antisense</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>Rosalind_3903_orf_23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Rosalind_3903_orf_24</td>\n",
       "      <td>234</td>\n",
       "      <td>102</td>\n",
       "      <td>Antisense +3</td>\n",
       "      <td>Antisense</td>\n",
       "      <td>132</td>\n",
       "      <td>44</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Rosalind_3903_orf_25</td>\n",
       "      <td>231</td>\n",
       "      <td>102</td>\n",
       "      <td>Antisense +3</td>\n",
       "      <td>Antisense</td>\n",
       "      <td>129</td>\n",
       "      <td>43</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Rosalind_3903_orf_26</td>\n",
       "      <td>123</td>\n",
       "      <td>102</td>\n",
       "      <td>Antisense +3</td>\n",
       "      <td>Antisense</td>\n",
       "      <td>21</td>\n",
       "      <td>7</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Rosalind_3903_orf_27</td>\n",
       "      <td>75</td>\n",
       "      <td>42</td>\n",
       "      <td>Antisense +3</td>\n",
       "      <td>Antisense</td>\n",
       "      <td>33</td>\n",
       "      <td>11</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Rosalind_3903_orf_28</td>\n",
       "      <td>37</td>\n",
       "      <td>16</td>\n",
       "      <td>Antisense +2</td>\n",
       "      <td>Antisense</td>\n",
       "      <td>21</td>\n",
       "      <td>7</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      ID  Start  Stop         Frame     Strand  Length  \\\n",
       "0    Rosalind_3903_orf_1     35    65      Sense +3      Sense      30   \n",
       "1    Rosalind_3903_orf_2    198   279      Sense +1      Sense      81   \n",
       "2    Rosalind_3903_orf_3    255   279      Sense +1      Sense      24   \n",
       "3    Rosalind_3903_orf_4    308   398      Sense +3      Sense      90   \n",
       "4    Rosalind_3903_orf_5    374   398      Sense +3      Sense      24   \n",
       "5    Rosalind_3903_orf_6    397   460      Sense +2      Sense      63   \n",
       "6    Rosalind_3903_orf_7    410   443      Sense +3      Sense      33   \n",
       "7    Rosalind_3903_orf_8    442   460      Sense +2      Sense      18   \n",
       "8    Rosalind_3903_orf_9    465   588      Sense +1      Sense     123   \n",
       "9   Rosalind_3903_orf_10    517   532      Sense +2      Sense      15   \n",
       "10  Rosalind_3903_orf_11    774   870      Sense +1      Sense      96   \n",
       "11  Rosalind_3903_orf_12    860   857  Antisense +1  Antisense       3   \n",
       "12  Rosalind_3903_orf_13    776   767  Antisense +1  Antisense       9   \n",
       "13  Rosalind_3903_orf_14    701   590  Antisense +1  Antisense     111   \n",
       "14  Rosalind_3903_orf_15    694   544  Antisense +2  Antisense     150   \n",
       "15  Rosalind_3903_orf_16    484   466  Antisense +2  Antisense      18   \n",
       "16  Rosalind_3903_orf_17    461   398  Antisense +1  Antisense      63   \n",
       "17  Rosalind_3903_orf_18    412   385  Antisense +2  Antisense      27   \n",
       "18  Rosalind_3903_orf_19    399   363  Antisense +3  Antisense      36   \n",
       "19  Rosalind_3903_orf_20    382   271  Antisense +2  Antisense     111   \n",
       "20  Rosalind_3903_orf_21    376   271  Antisense +2  Antisense     105   \n",
       "21  Rosalind_3903_orf_22    257   233  Antisense +1  Antisense      24   \n",
       "22  Rosalind_3903_orf_23    239   233  Antisense +1  Antisense       6   \n",
       "23  Rosalind_3903_orf_24    234   102  Antisense +3  Antisense     132   \n",
       "24  Rosalind_3903_orf_25    231   102  Antisense +3  Antisense     129   \n",
       "25  Rosalind_3903_orf_26    123   102  Antisense +3  Antisense      21   \n",
       "26  Rosalind_3903_orf_27     75    42  Antisense +3  Antisense      33   \n",
       "27  Rosalind_3903_orf_28     37    16  Antisense +2  Antisense      21   \n",
       "\n",
       "    Peptide size (aa)                                            Overlap  \n",
       "0                  10                                               None  \n",
       "1                  27  Rosalind_3903_orf_2, Rosalind_3903_orf_2, Rosa...  \n",
       "2                   8                                Rosalind_3903_orf_3  \n",
       "3                  30           Rosalind_3903_orf_4, Rosalind_3903_orf_4  \n",
       "4                   8           Rosalind_3903_orf_5, Rosalind_3903_orf_5  \n",
       "5                  21  Rosalind_3903_orf_6, Rosalind_3903_orf_6, Rosa...  \n",
       "6                  11           Rosalind_3903_orf_7, Rosalind_3903_orf_7  \n",
       "7                   6           Rosalind_3903_orf_8, Rosalind_3903_orf_8  \n",
       "8                  41           Rosalind_3903_orf_9, Rosalind_3903_orf_9  \n",
       "9                   5                               Rosalind_3903_orf_10  \n",
       "10                 32                               Rosalind_3903_orf_11  \n",
       "11                  1                               Rosalind_3903_orf_12  \n",
       "12                  3                                               None  \n",
       "13                 37                                               None  \n",
       "14                 50                                               None  \n",
       "15                  6                               Rosalind_3903_orf_16  \n",
       "16                 21                                               None  \n",
       "17                  9                                               None  \n",
       "18                 12                                               None  \n",
       "19                 37                                               None  \n",
       "20                 35                                               None  \n",
       "21                  8                               Rosalind_3903_orf_22  \n",
       "22                  2                               Rosalind_3903_orf_23  \n",
       "23                 44                                               None  \n",
       "24                 43                                               None  \n",
       "25                  7                                               None  \n",
       "26                 11                                               None  \n",
       "27                  7                                               None  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# Step 1: Prepare sequence for analysis\n",
    "with open('rosalind_orf.txt', 'r') as file:\n",
    "    infile = file.read().splitlines()\n",
    "    id = infile[0].replace('>','')\n",
    "    dna_seq = ''.join(infile[1:])\n",
    "    print(len(dna_seq))\n",
    "    rna_seq_sense = dna_seq.replace('T','U')\n",
    "    rna__seq_asense = rna_seq_sense.replace('A','u').replace('U','a').replace('C','g').replace('G','c').upper()[::-1]\n",
    "    rna_seq= {'Sense' : rna_seq_sense, 'Antisense' : rna__seq_asense}\n",
    "\n",
    "\n",
    "# Step 2: Set start and stop codons\n",
    "start_codon = ['AUG']\n",
    "stop_codon = ['UAA','UAG','UGA']\n",
    "\n",
    "\n",
    "# Step 3: Create each data variable as a list\n",
    "start = []\n",
    "stop = []\n",
    "strand = []\n",
    "frame = []\n",
    "length = []\n",
    "peptide_size = []\n",
    "overlap = []\n",
    "\n",
    "\n",
    "# Step 4: Fill-up the data lists\n",
    "for dir, seq in rna_seq.items():\n",
    "    for i in range(len(seq)-5): # -5 because I only want to find start codons that have enough subsequent sequence for a stop codon\n",
    "        if seq[i:i+3] in start_codon: # if you find a start codon...\n",
    "            for j in range(i+3 ,len(seq)-2, 3):\n",
    "                if seq[j:j+3] in stop_codon: # ...then find a stop codon, fill-out the variables: \n",
    "                    if dir == 'Sense':\n",
    "                        start.append(i)\n",
    "                        stop.append(j)\n",
    "                        strand.append(dir)\n",
    "                        if (i+3) % 3 == 0:\n",
    "                            frame.append(f'{dir} +1')\n",
    "                        elif (i+2) % 3 == 0:\n",
    "                            frame.append(f'{dir} +2')\n",
    "                        elif (i+1) % 3 == 0:\n",
    "                            frame.append(f'{dir} +3')\n",
    "                        length.append(j-i)\n",
    "                        peptide_size.append((j-i)//3) # // needed to get integer instead of float?\n",
    "                        break\n",
    "                    elif dir == 'Antisense':\n",
    "                        start.append(len(seq)-i)\n",
    "                        stop.append(len(seq)-j)\n",
    "                        strand.append(dir)\n",
    "                        if (i+3) % 3 == 0:\n",
    "                            frame.append(f'{dir} +1')\n",
    "                        elif (i+2) % 3 == 0:\n",
    "                            frame.append(f'{dir} +2')\n",
    "                        elif (i+1) % 3 == 0:\n",
    "                            frame.append(f'{dir} +3')\n",
    "                        length.append(j-i)\n",
    "                        peptide_size.append((j-i)//3)\n",
    "                        break\n",
    "\n",
    "\n",
    "orf_id = []\n",
    "for i in range(len(start)):\n",
    "    orf_id.append(f'{id}_orf_{i+1}')\n",
    "\n",
    "# This is kind of wrong because of the two directions.\n",
    "# So I won't be including it in the data fram until its fixed\n",
    "for k in range(len(orf_id)):\n",
    "    overlaps = []\n",
    "    for l in range(len(orf_id)):\n",
    "        if k != l:\n",
    "            if start[k] < stop[l] and stop[k] > start[l]:\n",
    "                overlaps.append(orf_id[k])\n",
    "    if overlaps:\n",
    "        overlap.append(', '.join(overlaps))\n",
    "    else:\n",
    "        overlap.append('None')\n",
    "\n",
    "\n",
    "# Step 5: Assemble data variable lists into a dictionary, then into a data-frame using pandas\n",
    "orf_data = {\n",
    "    'ID' : orf_id,\n",
    "    'Start' : start,\n",
    "    'Stop' : stop,\n",
    "    'Frame' : frame,\n",
    "    'Strand' : strand,\n",
    "    'Length' : length,\n",
    "    'Peptide size (aa)' : peptide_size\n",
    "    #'Overlap': overlap\n",
    "}\n",
    "\n",
    "import pandas as pd\n",
    "orf_df = pd.DataFrame(orf_data) # turn dictionary into data-frame\n",
    "orf_df.to_csv('orf_df.csv') # save in csv file\n",
    "\n",
    "orf_df\n",
    "\n",
    "# Missing:\n",
    "# Write csv file with dataframe\n",
    "# Write fasta file with peptides\n",
    "# Fix overlap situation\n",
    "\n",
    "\n",
    "#for index, row in df.iterrows():\n",
    "#    if row['Strand'] == 'Sense':\n",
    "#        print(rna_seq_sense[row['Start']:row['Stop']])\n",
    "#    elif row['Strand'] == 'Antisense':\n",
    "#        print(rna__seq_asense[row['Start']:row['Stop']])\n",
    "\n",
    "\n",
    "\n",
    "# Use Pandas to make dataframe with variables above\n",
    "# Then use it to select sense orfs and use start and stop to get DNA sequence, then to the same for antisense orfs\n",
    "# Then transcribe, then translate\n",
    "# Write fasta file for each sequence type\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2.2 Plot data\n",
    "\n",
    "Maybe at some point I will try to do it in python, but for now I wanted to do it in R.\n",
    "It would be a good way to warm up my R skills and get some analysis done before I dive into plotting with python.\n",
    "In the next block, I will try to do that R plot. Remember to change to the R kernel for it to work!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in library(tidyverse): there is no package called 'tidyverse'\n",
     "output_type": "error",
     "traceback": [
      "Error in library(tidyverse): there is no package called 'tidyverse'\nTraceback:\n",
      "1. library(tidyverse)"
     ]
    }
   ],
   "source": [
    "# Open csv file\n",
    "# Length of sequence is x axis\n",
    "# Putative ORF_id is the categorical y axis\n",
    "# Start and stop positions define size and position of bars plotted along the x axis\n",
    "# ORF_id are grouped depending on whether they occur in sense or antisense strand and placed, respectively, above or bellow the x axis\n",
    "\n",
    "library(tidyverse)\n",
    "library(ggplot2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error: package or namespace load failed for ‘ggplot2’ in loadNamespace(i, c(lib.loc, .libPaths()), versionCheck = vI[[i]]):\n namespace ‘scales’ 1.2.0 is being loaded, but >= 1.3.0 is required\n",
     "output_type": "error",
     "traceback": [
      "Error: package or namespace load failed for ‘ggplot2’ in loadNamespace(i, c(lib.loc, .libPaths()), versionCheck = vI[[i]]):\n namespace ‘scales’ 1.2.0 is being loaded, but >= 1.3.0 is required\nTraceback:\n",
      "1. library(ggplot2)",
      "2. tryCatch({\n .     attr(package, \"LibPath\") <- which.lib.loc\n .     ns <- loadNamespace(package, lib.loc)\n .     env <- attachNamespace(ns, pos = pos, deps, exclude, include.only)\n . }, error = function(e) {\n .     P <- if (!is.null(cc <- conditionCall(e))) \n .         paste(\" in\", deparse(cc)[1L])\n .     else \"\"\n .     msg <- gettextf(\"package or namespace load failed for %s%s:\\n %s\", \n .         sQuote(package), P, conditionMessage(e))\n .     if (logical.return && !quietly) \n .         message(paste(\"Error:\", msg), domain = NA)\n .     else stop(msg, call. = FALSE, domain = NA)\n . })",
      "3. tryCatchList(expr, classes, parentenv, handlers)",
      "4. tryCatchOne(expr, names, parentenv, handlers[[1L]])",
      "5. value[[3L]](cond)",
      "6. stop(msg, call. = FALSE, domain = NA)"
     ]
    }
   ],
   "source": [
    "# Plotting putative orf's on sequence\n",
    "\n",
    "# I actually think I want to do this with R ggplot\n",
    "#library(tidyverse)\n",
    "library(ggplot2)\n",
    "\n",
    "# Create a data frame with the categories, start, and stop positions\n",
    "df <- data.frame(\n",
    "  Category = c('A1', 'A2', 'B1', 'B2'),\n",
    "  Start = c(100, 400, 150, 500),\n",
    "  Stop = c(300, 700, 350, 800),\n",
    "  Group = c('Above', 'Above', 'Below', 'Below')\n",
    ")\n",
    "\n",
    "# Calculate the length of each bar\n",
    "df$Length <- df$Stop - df$Start\n",
    "\n",
    "# Create a new column for the adjusted position based on the group\n",
    "df$Position <- ifelse(df$Group == 'Above', df$Start, -df$Start)\n",
    "df$Length <- ifelse(df$Group == 'Above', df$Length, -df$Length)\n",
    "\n",
    "# Create the plot\n",
    "myplot <- ggplot(df, aes(x = Length, y = Category, fill = Group)) +\n",
    "  geom_bar(stat = \"identity\", aes(xmin = Position), width = 0.5) +\n",
    "  scale_x_continuous(limits = c(-1000, 1000)) +\n",
    "  scale_fill_manual(values = c('skyblue', 'lightgreen')) +\n",
    "  labs(\n",
    "    title = 'Horizontal Bars for Two Groups of Categories (Above and Below)',\n",
    "    x = 'Position', y = 'Categories'\n",
    "  ) +\n",
    "  theme_minimal()\n",
    "\n",
    "print(myplot)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
